hasher-matcher-actioner/hmalib/common/writebacker_models.py

```diff
@@ -0,0 +1,240 @@
+
+from hmalib.common.message_models import BankedSignal, MatchMessage
+from hmalib.common.logging import get_logger
+
+from hmalib.common.actioner_models import ActionPerformer
+from hmalib.common.label_models import (
+    WritebackLabel,
+    SawThisTooWritebackLabel,
+    IngestedWritebackLabel,
+    FalsePositiveWritebackLabel,
+    TruePositiveWritebackLabel,
+)
+from hmalib.common.actioner_models import ActionPerformer
+from hmalib.common.evaluator_models import ActionLabel
+from hmalib.common.message_models import WritebackMessage
+from hmalib.aws_secrets import AWSSecrets
+
+from threatexchange.api import ThreatExchangeAPI
+
+logger = get_logger(__name__)
+
+
+class MockedThreatExchangeAPI:
+    mocked_descriptor_ids = ["12345", "67890"]
+
+    def get_threat_descriptors_from_indicator(
+        self, indicator
+    ) -> t.List[t.Dict[str, str]]:
+        return [{"id": id} for id in self.mocked_descriptor_ids]
+
+    def react_to_threat_descriptor(self, descriptor, reaction) -> None:
+        assert descriptor in self.mocked_descriptor_ids
+
+
+class Writebacker:
+    """
+    For writing back to an HMA data soruce (eg. ThreatExchange). Every source that
+    enables writebacks should have an implmentation of this class
+    (eg ThreatExchangeWritebacker) and optionally sub implementations
+    (eg ThreatExchangeFalsePositiveWritebacker)
+
+    You must also add the subclass you are implementing to the performable_subclasses
+    fucntion below
+    """
+
+    @property
+    def source(self) -> str:
+        """
+        The source that this writebacker corresponds to (eg. "te")
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    def writeback_options() -> t.Dict[str, t.Type["Writebacker"]]:
+        """
+        For a given source that performs writebacks, this fucntion specifies what types of
+        writebacks that can be taken as a mapping from writback type to writebacker. The
+        type should be same as the type of is passed in the WritebackLabel passed to the writebacker
+        """
+        raise NotImplementedError
+
+    @classmethod
+    @lru_cache(maxsize=None)
+    def get_writebacker_for_source(cls, source: str) -> t.Optional["Writebacker"]:
+        if cls.__name__ != "Writebacker":
+            raise ValueError(
+                "get_writebacker_for_source can only be called from the Writebacker class directly. eg Writebacker().get_writebacker_for_source"
+            )
+
+        print("subclasses", cls.__subclasses__())
+
+        sources_to_writebacker_cls = {
+            writebacker_cls().source: writebacker_cls
+            for writebacker_cls in cls.__subclasses__()
+        }
+
+        if source not in sources_to_writebacker_cls.keys():
+            return None
+        return sources_to_writebacker_cls[source]()
+
+    def writeback_is_enabled(self) -> bool:
+        """
+        Users can switch on/off writebacks either globally or for individual sources
+        """
```

Comment:
Blocking: I thought that we enable writebacks per "dataset" ("collaboration")? It's not at the source level. Mengyang is adding a field in the dataset config.